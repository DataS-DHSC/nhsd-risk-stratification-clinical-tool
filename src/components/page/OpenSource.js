import { Details } from 'nhsuk-react-components';
import React from 'react';
import PropTypes from 'prop-types';
import PageReviewDates from '../PageReviewDates';
import SiteBrandingTitle from '../SiteBrandingTitle';

const OpenSource = ({ licenseMapArray }) => {
  const licenseWithWarning = ['GPL3', 'Unknown'];

  const dependencies = licenseMapArray.flatMap((licenseMap) =>
    Object.entries(licenseMap).map(([key, value]) => [
      value.licenses ?? 'Unknown',
      key,
    ])
  );
  const licenses = [...new Set(dependencies.map((d) => d[0]))].sort();

  const openSourceSummaries = licenses.map((license, index) => (
    <Details id={`nhsuk-details${index}`} key={`nhsuk-details-${license}`}>
      <Details.Summary
        id={`nhsuk-details__summary${index}`}
        role="button"
        tabIndex="0"
        aria-expanded="true"
        className={
          licenseWithWarning.includes(license) ? 'warning-license' : ''
        }
        aria-controls={`nhsuk-details__text${index}`}
      >
        {license}
      </Details.Summary>
      <Details.Text id={`nhsuk-details__text${index}`}>
        <p id={`nhsuk-details__content${index}`}>
          {dependencies
            .filter((dependency) => dependency[0] === license)
            .map((d) => d[1])
            .join(', ')}
        </p>
      </Details.Text>
    </Details>
  ));

  return (
    <>
      <div className="nhsuk-grid-row">
        <div className="nhsuk-grid-column-two-thirds">
          <h1>
            Open-source licenses for the <SiteBrandingTitle isDefaultName />
          </h1>
          {openSourceSummaries}
          <a
            className="nhsuk-body-m license-definitions"
            href="https://spdx.org/licenses/"
            target="_self"
            rel="noopener noreferrer nofollow"
          >
            License definitions.
          </a>
        </div>
      </div>
      <PageReviewDates page="static" />
    </>
  );
};

OpenSource.propTypes = {
  licenseMapArray: PropTypes.arrayOf(PropTypes.shape()).isRequired,
};

export default OpenSource;
