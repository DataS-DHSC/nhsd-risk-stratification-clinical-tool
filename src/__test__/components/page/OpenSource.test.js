import React from 'react';
import { render } from '@testing-library/react';
import OpenSource from '../../../components/page/OpenSource';

const inputLicenses1 = {
  '@apidevtools/json-schema-ref-parser@9.0.7': {
    licenses: 'MIT',
    repository: 'https://github.com/APIDevTools/json-schema-ref-parser',
    publisher: 'James Messinger',
    url: 'https://jamesmessinger.com',
    name: '@apidevtools/json-schema-ref-parser',
    version: '9.0.7',
    description: 'Parse, Resolve, and Dereference JSON Schema $ref pointers',
    licenseFile: 'node_modules/@apidevtools/json-schema-ref-parser/LICENSE',
    copyright: 'Copyright (c) 2015 James Messinger',
  },
  '@aws-crypto/sha256-browser@1.1.0': {
    licenses: 'Apache-2.0',
    repository: 'https://github.com/aws/aws-sdk-js-crypto-helpers',
    publisher: 'AWS Crypto Tools Team',
    email: 'aws-cryptools@amazon.com',
    url:
      'https://docs.aws.amazon.com/aws-crypto-tools/index.html?id=docs_gateway#lang/en_us',
    name: '@aws-crypto/sha256-browser',
    version: '1.1.0',
    description:
      'SHA256 wrapper for browsers that prefers `window.crypto.subtle` but will fall back to a pure JS implementation in @aws-crypto/sha256-js to provide a consistent interface for SHA256.',
    licenseFile: 'node_modules/@aws-crypto/sha256-browser/LICENSE',
  },
  '@aws-crypto/sha256-js@1.1.0': {
    licenses: 'Apache-2.0',
    repository: 'https://github.com/aws/aws-sdk-js-crypto-helpers',
    publisher: 'AWS Crypto Tools Team',
    email: 'aws-cryptools@amazon.com',
    url:
      'https://docs.aws.amazon.com/aws-crypto-tools/index.html?id=docs_gateway#lang/en_us',
    name: '@aws-crypto/sha256-js',
    version: '1.1.0',
    description: 'A pure JS implementation SHA256.',
    licenseFile: 'node_modules/@aws-crypto/sha256-js/LICENSE',
  },
  'jest-environment-jsdom@26.6.2': {
    licenses: 'MIT',
    repository: 'https://github.com/facebook/jest',
    name: 'jest-environment-jsdom',
    version: '26.6.2',
    description: '',
    licenseFile: 'node_modules/jest-environment-jsdom/LICENSE',
    copyright: 'Copyright (c) Facebook, Inc. and its affiliates.',
  },
  'jest-environment-node@26.6.2': {
    licenses: 'MIT',
    repository: 'https://github.com/facebook/jest',
    name: 'jest-environment-node',
    version: '26.6.2',
    description: '',
    licenseFile: 'node_modules/jest-environment-node/LICENSE',
    copyright: 'Copyright (c) Facebook, Inc. and its affiliates.',
  },
  'webdriver@7.5.3': {
    licenses: 'MIT',
    repository: 'https://github.com/webdriverio/webdriverio',
    publisher: 'Christian Bromann',
    email: 'christian@saucelabs.com',
    name: 'webdriver',
    version: '7.5.3',
    description:
      'A Node.js bindings implementation for the W3C WebDriver and Mobile JSONWire Protocol',
    licenseFile: 'node_modules/webdriver/LICENSE-MIT',
    copyright: 'Copyright (c) OpenJS Foundation and other contributors',
  },
};

const inputLicenses2 = {
  MITFluentAssertions: {
    licenses: 'Apache-2.0',
  },
  'coverlet.collector': {
    licenses: 'MIT',
  },
  AWSXRayRecorder: {
    licenses: 'Apache-2.0',
  },
  'Amazon.Lambda.Core': {
    licenses: 'Apache-2.0',
  },
};

const setup = () =>
  render(<OpenSource licenseMapArray={[inputLicenses1, inputLicenses2]} />);

describe('OpenSource', () => {
  it('should match snapshot', () => {
    const container = setup();

    expect(container.asFragment()).toMatchSnapshot();
  });
});
