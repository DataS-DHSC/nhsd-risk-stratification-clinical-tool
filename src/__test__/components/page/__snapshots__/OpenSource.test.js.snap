// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenSource should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-grid-row"
  >
    <div
      class="nhsuk-grid-column-two-thirds"
    >
      <h1>
        Open-source licenses for the COVID-19 Clinical Risk Assessment Tool
      </h1>
      <details
        class="nhsuk-details"
        id="nhsuk-details0"
      >
        <summary
          aria-controls="nhsuk-details__text0"
          aria-expanded="true"
          class="nhsuk-details__summary"
          id="nhsuk-details__summary0"
          role="button"
          tabindex="0"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            Apache-2.0
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          id="nhsuk-details__text0"
        >
          <p
            id="nhsuk-details__content0"
          >
            @aws-crypto/sha256-browser@1.1.0, @aws-crypto/sha256-js@1.1.0, MITFluentAssertions, AWSXRayRecorder, Amazon.Lambda.Core
          </p>
        </div>
      </details>
      <details
        class="nhsuk-details"
        id="nhsuk-details1"
      >
        <summary
          aria-controls="nhsuk-details__text1"
          aria-expanded="true"
          class="nhsuk-details__summary"
          id="nhsuk-details__summary1"
          role="button"
          tabindex="0"
        >
          <span
            class="nhsuk-details__summary-text"
          >
            MIT
          </span>
        </summary>
        <div
          class="nhsuk-details__text"
          id="nhsuk-details__text1"
        >
          <p
            id="nhsuk-details__content1"
          >
            @apidevtools/json-schema-ref-parser@9.0.7, jest-environment-jsdom@26.6.2, jest-environment-node@26.6.2, webdriver@7.5.3, coverlet.collector
          </p>
        </div>
      </details>
      <a
        class="nhsuk-body-m license-definitions"
        href="https://spdx.org/licenses/"
        rel="noopener noreferrer nofollow"
        target="_self"
      >
        License definitions.
      </a>
    </div>
  </div>
  <div
    class="nhsuk-review-date"
    id="review-date"
  >
    <p
      class="nhsuk-body-s"
    >
      Page last reviewed: 1 December 2020
      <br />
      Next review due: 23 December 2023
    </p>
  </div>
</DocumentFragment>
`;
